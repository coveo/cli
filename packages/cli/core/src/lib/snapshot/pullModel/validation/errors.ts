import {ValidationError} from 'jsonschema';
import {CLIBaseError} from '@coveo/cli-commons/errors/cliBaseError';
import {getPrettyJsonValidationErrors} from '@coveo/cli-commons/errors/jsonError';

abstract class BaseSPMError extends CLIBaseError {
  public name = 'Snapshot Pull Model Error';
  private static messageSuffix =
    '\nThis is probably a problem with the Coveo CLI, please report this issue at https://github.com/coveo/cli/issues';
  public constructor(message: string, shouldContactCoveo: boolean) {
    super(`${message}${shouldContactCoveo ? BaseSPMError.messageSuffix : ''}`);
  }
}

export class UnknownSPMValidationError extends BaseSPMError {
  public name = 'Unknown Snapshot Pull Model Error';
  public constructor(shouldContactCoveo: boolean) {
    super(
      "An unknown error occured while validating the custom template. Try recreating it from 'empty' or 'full'.",
      shouldContactCoveo
    );
  }
}

export class InvalidSPMError extends BaseSPMError {
  public name = 'Invalid Snapshot Pull Model Error';
  public constructor(shouldContactCoveo: boolean, errors: ValidationError[]) {
    super(getPrettyJsonValidationErrors(errors), shouldContactCoveo);
  }
}
