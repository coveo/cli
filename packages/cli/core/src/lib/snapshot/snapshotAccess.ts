import PlatformClient, {
  ResourceSnapshotType,
  SnapshotAccessType,
} from '@coveo/platform-client';
import {
  MissingResourcePrivileges,
  MissingSnapshotPrivilege,
} from '../errors/snapshotErrors';
import {isSubset, without} from '../utils/list';

export async function ensureResourcesAccess(
  client: PlatformClient,
  resourceTypes: ResourceSnapshotType[],
  snapshotAccessType: SnapshotAccessType = SnapshotAccessType.Read
) {
  const allowedResources = await client.resourceSnapshot.listResourceAccess();
  const allowed = isSubset(resourceTypes, allowedResources);
  if (!allowed) {
    const missingResources = without(resourceTypes, allowedResources);
    throw new MissingResourcePrivileges(missingResources, snapshotAccessType);
  }
}

export async function ensureSnapshotAccess(
  client: PlatformClient,
  snapshotId: string,
  snapshotAccessType: SnapshotAccessType = SnapshotAccessType.Read
) {
  const {allowed} = await client.resourceSnapshot.validateAccess(snapshotId, {
    snapshotAccessType,
  });

  if (!allowed) {
    throw new MissingSnapshotPrivilege(snapshotId, snapshotAccessType);
  }
}
