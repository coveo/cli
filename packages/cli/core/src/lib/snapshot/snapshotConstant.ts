import {ResourceSnapshotType} from '@coveo/platform-client';

const snapshotLabelDict = {
  [ResourceSnapshotType.extension]: 'Extensions',
  [ResourceSnapshotType.featuredResult]: 'Featured result rules',
  [ResourceSnapshotType.field]: 'Fields',
  [ResourceSnapshotType.filter]: 'Filter rules',
  [ResourceSnapshotType.mapping]: 'Source mappings',
  [ResourceSnapshotType.mlModel]: 'Machine learning models',
  [ResourceSnapshotType.mlModelAssociation]:
    'Machine learning model associations',
  [ResourceSnapshotType.queryParameter]: 'Query parameter rules',
  [ResourceSnapshotType.queryPipeline]: 'Query pipelines',
  [ResourceSnapshotType.queryPipelineCondition]: 'Conditions',
  [ResourceSnapshotType.rankingExpression]: 'Ranking expression rules',
  [ResourceSnapshotType.rankingWeight]: 'Ranking weight rules',
  [ResourceSnapshotType.searchPage]: 'Search pages',
  [ResourceSnapshotType.securityProvider]: 'Security providers',
  [ResourceSnapshotType.source]: 'Sources',
  [ResourceSnapshotType.stopWord]: 'Stop word rules',
  [ResourceSnapshotType.subscription]: 'Notifications',
  [ResourceSnapshotType.thesaurus]: 'Thesaurus rules',
  [ResourceSnapshotType.trigger]: 'Trigger rules',
};

export const disabledResourceType = [
  ResourceSnapshotType.mlModel,
  ResourceSnapshotType.mlModelAssociation,
];
export const allowedResourceType = Object.values(ResourceSnapshotType).filter(
  (resourceType) => !disabledResourceType.includes(resourceType)
);

type ExtensibleSnapshotLabelDict = typeof snapshotLabelDict &
  Record<string, string>;

export const labels: ExtensibleSnapshotLabelDict = snapshotLabelDict;
