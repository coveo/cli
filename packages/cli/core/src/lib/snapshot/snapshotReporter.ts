import {
  ResourceSnapshotsReportModel,
  ResourceSnapshotsReportOperationModel,
  ResourceSnapshotsReportOperationResult,
  ResourceSnapshotType,
} from '@coveo/platform-client';
import {
  ReportViewerOperationName,
  ReportViewerResourceReportModel,
  SnapshotReportStatus,
} from './reportPreviewer/reportPreviewerDataModels';

type ResourceEntries = [string, ResourceSnapshotsReportOperationModel];

type NoopHandler = (this: void) => void;
type SnapshotReporterHandler = (this: SnapshotReporter) => void | Promise<void>;

type SnapshotReporterHandlers = {
  [K in SnapshotReportStatus]: SnapshotReporterHandler | NoopHandler;
};

const MissingVaultEntryResultCode = 'INVALID_PLACEHOLDER';
//#endregion

export type VaultEntryAttributes = {
  vaultEntryId: string;
  resourceName: string;
  resourceType: ResourceSnapshotType;
};

export class SnapshotReporter {
  private missingVaultEntriesSet: Set<VaultEntryAttributes> = new Set();

  private operationResultsParseFuseUsed: boolean = false;

  private consumeOperationResultsParseFuse() {
    if (this.operationResultsParseFuseUsed) {
      return false;
    } else {
      this.operationResultsParseFuseUsed = true;
      return true;
    }
  }

  public get missingVaultEntries() {
    return this.missingVaultEntriesSet.values();
  }
  public resourceInErrorCount: number = 0;
  public resourceInError: Map<Partial<ResourceSnapshotType>, Set<string>> =
    new Map();
  public constructor(public readonly report: ResourceSnapshotsReportModel) {}

  public static convertResourceEntriesToResourceSnapshotsReportOperationModel([
    name,
    operations,
  ]: ResourceEntries) {
    return {name, operations};
  }

  public static resourceHasAtLeastOneOperationFilter(
    operationsToDisplay: ReportViewerOperationName[]
  ) {
    return ([_, operations]: ResourceEntries) =>
      operationsToDisplay.reduce(
        (previous, current) => previous + operations[current],
        0
      ) > 0;
  }

  public getAffectedResourceTypes(): ResourceSnapshotType[] {
    const isResourceTypeReportAffected = (
      report: ResourceSnapshotsReportOperationModel
    ) => Object.values(report).some((operation) => operation > 0);
    const changedResourcesTypes: ResourceSnapshotType[] = [];
    for (const resourceType of Object.keys(this.report.resourceOperations)) {
      if (
        isResourceTypeReportAffected(
          this.report.resourceOperations[resourceType]
        )
      ) {
        changedResourcesTypes.push(resourceType as ResourceSnapshotType);
      }
    }
    return changedResourcesTypes;
  }

  public hasChangedResources() {
    const totalUnchanges =
      this.getOperationTypeTotalCount('resourcesUnchanged');
    return totalUnchanges !== this.report.resourcesProcessed;
  }

  public getOperationTypeTotalCount(
    type: keyof ResourceSnapshotsReportOperationModel
  ) {
    const count = Object.values(this.report.resourceOperations).reduce(
      (count: number, current: ResourceSnapshotsReportOperationModel) =>
        count + current[type],
      0
    );

    return count;
  }

  private reportHandlers: SnapshotReporterHandlers = {
    [SnapshotReportStatus.SUCCESS]: () => {},
    [SnapshotReportStatus.NO_CHANGES]: () => {},
    [SnapshotReportStatus.MISSING_VAULT_ENTRIES]: () => {},
    [SnapshotReportStatus.ERROR]: () => {},
  };

  public setReportHandler<T extends SnapshotReportStatus>(
    status: T,
    handler: SnapshotReporterHandlers[T]
  ): SnapshotReporter {
    this.reportHandlers[status] = handler;
    return this;
  }

  public async handleReport(): Promise<void> {
    const reportStatuses = this.getReportStatuses();
    await this.executeHandlers(reportStatuses.fixables);
    if (reportStatuses.errors.length > 0) {
      await this.executeHandlers(reportStatuses.errors);
      return;
    }
    if (reportStatuses.successes.length > 0) {
      await this.executeHandlers(reportStatuses.successes);
    }
  }

  private async executeHandlers(reportStatuses: SnapshotReportStatus[]) {
    for (const status of reportStatuses) {
      await this.callAndReset(status);
    }
  }

  private async callAndReset(handler: SnapshotReportStatus) {
    await this.reportHandlers[handler].apply(this);
    this.reportHandlers[handler] = () => {};
  }

  private getReportStatuses() {
    const statuses: Record<
      'fixables' | 'successes' | 'errors',
      SnapshotReportStatus[]
    > = {
      fixables: [],
      successes: [],
      errors: [],
    };

    this.parseResourcesOperationsResults();
    if (this.missingVaultEntriesSet.size > 0) {
      statuses.fixables.push(SnapshotReportStatus.MISSING_VAULT_ENTRIES);
    }

    if (!this.hasChangedResources()) {
      statuses.successes.push(SnapshotReportStatus.NO_CHANGES);
    }

    if (this.resourceInErrorCount > 0) {
      statuses.errors.push(SnapshotReportStatus.ERROR);
    } else {
      statuses.successes.push(SnapshotReportStatus.SUCCESS);
    }

    return statuses;
  }

  private parseResourcesOperationsResults(): void {
    if (!this.consumeOperationResultsParseFuse()) {
      return;
    }
    for (const [resourceType, resource] of Object.entries(
      this.report.resourceOperationResults
    )) {
      for (const [resourceName, errors] of Object.entries(resource)) {
        for (const err of errors) {
          this.parseResourceOperationResult(err, resourceName, resourceType);
        }
      }
    }
  }

  private parseResourceOperationResult(
    err: ResourceSnapshotsReportOperationResult,
    resourceName: string,
    resourceType: string
  ) {
    const missingEntry = SnapshotReporter.tryGetMissingVaultEntryName(err);
    if (missingEntry) {
      this.missingVaultEntriesSet.add({
        vaultEntryId: missingEntry,
        resourceName: resourceName,
        resourceType: resourceType as ResourceSnapshotType,
      });
    } else {
      // TODO: Fix PlatformClient to reflect proper typing.
      this.addResourceInError(resourceType as ResourceSnapshotType, err);
    }
  }

  private addResourceInError(
    resourceType: ResourceSnapshotType,
    err: ResourceSnapshotsReportOperationResult
  ) {
    this.resourceInErrorCount++;
    let errorSet = this.resourceInError.get(resourceType);
    if (!errorSet) {
      errorSet = new Set();
      this.resourceInError.set(resourceType, errorSet);
    }
    errorSet.add(`${err.resultCode}: ${err.message}`);
  }

  private static missingVaultEntryMatcher =
    /^The vault entry referenced by \{\{ VAULT\.(?<entryName>.*) \}\} could not be found in the vault\.$/;

  private static tryGetMissingVaultEntryName(
    err: ResourceSnapshotsReportOperationResult
  ): string | undefined {
    let message: string = '';
    if (typeof err !== 'string') {
      if (err.resultCode === MissingVaultEntryResultCode) {
        message = err.message;
      } else {
        return undefined;
      }
    } else {
      message = err;
    }
    const match = message.match(SnapshotReporter.missingVaultEntryMatcher);
    return match?.groups?.['entryName'];
  }

  public getChangedResources(
    operationsToDisplay: ReportViewerOperationName[]
  ): ReportViewerResourceReportModel[] {
    return Object.entries(this.report.resourceOperations)
      .filter(
        SnapshotReporter.resourceHasAtLeastOneOperationFilter(
          operationsToDisplay
        )
      )
      .map(
        SnapshotReporter.convertResourceEntriesToResourceSnapshotsReportOperationModel
      );
  }

  public get type() {
    const type = this.prettify(this.report.type);
    return type;
  }

  public get status() {
    const status = this.prettify(this.report.status);
    return status;
  }

  public get resultCode() {
    const status = this.prettify(this.report.resultCode);
    return status;
  }

  private prettify(str: string): string {
    const capitalized = (word: string) =>
      word.charAt(0) + word.slice(1).toLowerCase();

    const spaced = str.replace(/_/g, ' ');
    return spaced.split(' ').map(capitalized).join(' ');
  }
}
