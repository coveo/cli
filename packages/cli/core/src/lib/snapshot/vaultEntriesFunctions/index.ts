export * from './interfaces.js';
export * from './transferFromOrganization.js';
export * from './tryCreateMissingVaultEntries.js';
export * from './throwSnapshotMissingVaultEntriesError.js';
