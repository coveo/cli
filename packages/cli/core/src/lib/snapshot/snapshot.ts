import {
  ResourceSnapshotsModel,
  ResourceSnapshotsReportModel,
  ResourceSnapshotsReportResultCode,
  ResourceSnapshotsReportStatus,
  PlatformClient,
  ResourceSnapshotsReportType,
  SnapshotExportContentFormat,
  ApplyOptionsDeletionScope,
} from '@coveo/platform-client';
import retry from 'async-retry';
import {ReportViewer} from './reportPreviewer/reportPreviewer';
import {ensureFileSync, writeJsonSync} from 'fs-extra';
import {join} from 'path';
import {SnapshotReporter} from './snapshotReporter';
import {SnapshotOperationTimeoutError} from '../errors';
import {ExpandedPreviewer} from './expandedPreviewer/expandedPreviewer';
import {Project} from '../project/project';
import {SnapshotNoReportFoundError} from '../errors/snapshotErrors';
import {SnapshotReportStatus} from './reportPreviewer/reportPreviewerDataModels';

export type SnapshotReport = ResourceSnapshotsReportModel;

export interface WaitUntilDoneOptions {
  /**
   * The maximum number of seconds to wait before the commands exits with a timeout error.
   * A value of zero will prevent the command from timing out.
   */
  wait?: number; // in seconds
  /**
   * The interval between 2 consecutive polls.
   */
  waitInterval?: number; // in seconds
  /**
   * Callback to execute every time a request is being made to retrieve the snapshot data
   */
  onRetryCb?: (report: ResourceSnapshotsReportModel) => void;
}

export interface WaitUntilOperationDone extends WaitUntilDoneOptions {
  /**
   * The operation to wait for. If not specified, the method will wait for any operation to complete.
   */
  operationToWaitFor?: ResourceSnapshotsReportType;
}

export class Snapshot {
  public static defaultWaitOptions: Required<WaitUntilDoneOptions> = {
    waitInterval: 1,
    wait: 60,
    onRetryCb: (_report: ResourceSnapshotsReportModel) => {},
  };

  private static ongoingReportStatuses = [
    ResourceSnapshotsReportStatus.Pending,
    ResourceSnapshotsReportStatus.InProgress,
  ];

  public constructor(
    private model: ResourceSnapshotsModel,
    private client: PlatformClient
  ) {}

  public async validate(
    deleteMissingResources = false,
    options: WaitUntilDoneOptions = {}
  ): Promise<SnapshotReporter> {
    await this.snapshotClient.dryRun(this.id, {
      deleteMissingResources,
    });

    await this.waitUntilDone({
      operationToWaitFor: ResourceSnapshotsReportType.DryRun,
      ...options,
    });

    return new SnapshotReporter(this.latestReport);
  }

  public async preview(
    projectToPreview: Project,
    deleteMissingResources = false,
    expandedPreview = true
  ) {
    const reporter = new SnapshotReporter(this.latestReport);
    await this.displayLightPreview(reporter);
    if (!expandedPreview) {
      return;
    }
    const onSuccess = () =>
      this.generateExpandedPreview(projectToPreview, deleteMissingResources);
    await reporter
      .setReportHandler(
        SnapshotReportStatus.SUCCESS,
        async function (this: SnapshotReporter) {
          await onSuccess();
        }
      )
      .handleReport();
  }

  public async apply(
    deleteMissingResources = false,
    options: WaitUntilDoneOptions = {}
  ) {
    await this.snapshotClient.apply(
      this.id,
      deleteMissingResources
        ? {
            deleteMissingResources,
            deletionScope: ApplyOptionsDeletionScope.OnlyTypesFromSnapshot,
          }
        : {deleteMissingResources}
    );

    await this.waitUntilDone({
      operationToWaitFor: ResourceSnapshotsReportType.Apply,
      ...options,
    });

    return new SnapshotReporter(this.latestReport);
  }

  public async delete() {
    await this.client.resourceSnapshot.delete(this.model.id);
  }

  public download() {
    return this.client.resourceSnapshot.export(this.id, {
      contentFormat: SnapshotExportContentFormat.SplitPerType,
    });
  }

  public areResourcesInError() {
    return (
      this.latestReport.resultCode ===
      ResourceSnapshotsReportResultCode.ResourcesInError
    );
  }

  public saveDetailedReport(projectPath: string) {
    const pathToReport = join(
      projectPath,
      'snapshot-reports',
      `${this.latestReport.id}.json`
    );
    ensureFileSync(pathToReport);
    writeJsonSync(pathToReport, this.latestReport, {spaces: 2});
    return pathToReport;
  }

  public get latestReport() {
    const reports = this.model.reports;
    if (!Array.isArray(reports) || reports.length === 0) {
      throw new SnapshotNoReportFoundError(this);
    }
    return this.sortReportsByDate(reports)[0];
  }

  public get id() {
    return this.model.id;
  }

  public get targetId() {
    return this.model.targetId;
  }

  private get snapshotClient() {
    return this.client.resourceSnapshot;
  }

  private sortReportsByDate<T extends SnapshotReport>(report: T[]): T[] {
    return report.sort((a, b) => b.updatedDate - a.updatedDate);
  }

  private async displayLightPreview(reporter: SnapshotReporter) {
    const viewer = new ReportViewer(reporter);
    await viewer.display();
  }

  private async generateExpandedPreview(
    projectToPreview: Project,
    shouldDelete: boolean
  ) {
    const previewer = new ExpandedPreviewer(
      this.latestReport,
      this.targetId,
      projectToPreview,
      shouldDelete
    );
    await previewer.preview();
  }

  private async refreshSnapshotData() {
    this.model = await this.snapshotClient.get(this.model.id, {
      includeReports: true,
    });
  }

  public waitUntilDone(options: WaitUntilOperationDone = {}) {
    const opts = {...Snapshot.defaultWaitOptions, ...options};
    const toMilliseconds = (seconds: number) => seconds * 1e3;

    return retry(
      this.waitUntilDoneRetryFunction(opts.onRetryCb, opts.operationToWaitFor),
      // Setting the retry mechanism to follow a time-based logic instead of specifying the  number of attempts.
      {
        retries: Math.ceil(opts.wait / opts.waitInterval),
        forever: opts.wait === 0,
        minTimeout: toMilliseconds(opts.waitInterval),
        maxTimeout: toMilliseconds(opts.waitInterval),
        maxRetryTime: toMilliseconds(opts.wait),
      }
    );
  }

  private isUnsettled() {
    return Snapshot.ongoingReportStatuses.includes(this.latestReport.status);
  }

  private isGoingThroughOperation(
    operation: ResourceSnapshotsReportType
  ): boolean {
    if (this.latestReport.type === operation) {
      return true;
    }
    return false;
  }

  private waitUntilDoneRetryFunction(
    onRetryCb: (report: ResourceSnapshotsReportModel) => void,
    operationToWaitFor?: ResourceSnapshotsReportType
  ): () => Promise<void> {
    return (async () => {
      await this.refreshSnapshotData();

      const isUnsettled = this.isUnsettled();
      const isUnexpectedOperation = operationToWaitFor
        ? !this.isGoingThroughOperation(operationToWaitFor)
        : false;

      onRetryCb(this.latestReport);

      if (isUnsettled || isUnexpectedOperation) {
        throw new SnapshotOperationTimeoutError(this);
      }
    }).bind(this);
  }
}
