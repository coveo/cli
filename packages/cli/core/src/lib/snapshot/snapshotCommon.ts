import {startSpinner} from '@coveo/cli-commons/utils/ux';
import {Project} from '../project/project';
import {SnapshotFactory} from './snapshotFactory';
import {Snapshot, WaitUntilDoneOptions} from './snapshot';
import {normalize} from 'path';
import {Config, Configuration} from '@coveo/cli-commons/config/config';
import {SnapshotGenericError} from '../errors/snapshotErrors';
import {SnapshotReporter} from './snapshotReporter';
import {
  VaultTransferFunctionsParam,
  tryTransferFromOrganization,
  throwSnapshotMissingVaultEntriesError,
  tryCreateMissingVaultEntries,
} from './vaultEntriesFunctions';

export interface DryRunOptions {
  deleteMissingResources?: boolean;
  snapshotId?: string;
  waitUntilDone?: WaitUntilDoneOptions;
}

async function internalDryRun(snapshot: Snapshot, options: DryRunOptions) {
  return snapshot.validate(
    options.deleteMissingResources,
    options.waitUntilDone
  );
}

export async function dryRun(
  targetOrg: string,
  projectPath: string,
  options: DryRunOptions = {}
) {
  const project = new Project(normalize(projectPath), targetOrg);
  const snapshot = await getSnapshotForDryRun(project, targetOrg, options);

  startSpinner('Validating snapshot');
  let reporter = await internalDryRun(snapshot, options);

  return {reporter, snapshot, project};
}

export function getTargetOrg(config: Config, target?: string) {
  if (target) {
    return target;
  }
  const cfg = config.get();
  return cfg.organization;
}

export function cleanupProject(projectPath: string) {
  const project = new Project(normalize(projectPath));
  project.deleteTemporaryZipFile();
}

export function getErrorReportHandler(
  snapshot: Snapshot,
  cfg: Configuration,
  projectPath?: string
) {
  return () => {
    throw new SnapshotGenericError(snapshot, cfg, projectPath);
  };
}

export async function handleReportWithErrors(
  snapshot: Snapshot,
  cfg: Configuration,
  projectPath?: string
) {
  throw new SnapshotGenericError(snapshot, cfg, projectPath);
}

type MissingVaultEntriesReportSubHandler = (
  param: VaultTransferFunctionsParam
) => Promise<boolean> | Promise<never> | boolean | never;

const missingVaultEntriesSubHandlers: MissingVaultEntriesReportSubHandler[] = [
  tryTransferFromOrganization,
  tryCreateMissingVaultEntries,
  throwSnapshotMissingVaultEntriesError,
];

export function getMissingVaultEntriesReportHandler(
  snapshot: Snapshot,
  cfg: Configuration,
  projectPath?: string
) {
  return async function (this: SnapshotReporter) {
    for (const subHandler of missingVaultEntriesSubHandlers) {
      if (await subHandler({reporter: this, snapshot, cfg, projectPath})) {
        break;
      }
    }
  };
}

async function getSnapshotForDryRun(
  project: Project,
  targetOrg: string,
  options: DryRunOptions = {}
) {
  if (options.snapshotId) {
    startSpinner('Retrieving Snapshot');
    return SnapshotFactory.createFromExistingSnapshot(
      options.snapshotId,
      targetOrg,
      options.waitUntilDone
    );
  }
  startSpinner('Creating Snapshot');
  return SnapshotFactory.createSnapshotFromProject(
    project,
    targetOrg,
    options.waitUntilDone
  );
}
