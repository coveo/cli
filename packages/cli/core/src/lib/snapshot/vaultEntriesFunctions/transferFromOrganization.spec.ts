jest.mock('@coveo/cli-commons/lib/platform/authenticatedClient');
jest.mock('../../project/project');
jest.mock('../snapshot');
jest.mock('../snapshotReporter');
jest.mock('@oclif/core/lib/cli-ux/prompt');
jest.mock('@oclif/core/lib/errors/index');
jest.mock('../../errors/vaultErrors');

import PlatformClient, {VaultFetchStrategy} from '@coveord/platform-client';
import {confirm} from '@oclif/core/lib/cli-ux/prompt';
import {warn} from '@oclif/core/lib/errors/index';
import {Configuration} from '@coveo/cli-commons/lib/config/config';
import {SnapshotMissingVaultEntriesFromOriginError} from '../../errors/vaultErrors';
import {AuthenticatedClient} from '@coveo/cli-commons/lib/platform/authenticatedClient';
import {Project} from '../../project/project';
import {Snapshot} from '../snapshot';
import {SnapshotReporter} from '../snapshotReporter';
import {tryTransferFromOrganization} from './transferFromOrganization';

describe('#tryTransferFromOrganization', () => {
  const mockedProject = jest.mocked(Project);
  const mockedSnapshotMissingVaultEntriesFromOriginError = jest.mocked(
    SnapshotMissingVaultEntriesFromOriginError
  );
  const mockedCliConfirm = jest.mocked(confirm);
  const mockedCliWarn = jest.mocked(warn);
  const mockedVaultList = jest.fn();
  const mockedVaultImport = jest.fn();
  const mockedGetClient = jest.fn();
  const mockedGetUserHasAccessToOrg = jest.fn();
  const mockedAuthenticatedClient = jest.mocked(AuthenticatedClient);

  const doMockProjectWithOriginOrg = () => {
    mockedProject.mockImplementation(
      () =>
        ({
          getResourceManifest: () => ({orgId: 'someOriginOrg'}),
        } as Partial<Project> as Project)
    );
  };

  const doMockUserAnswerYes = () => {
    mockedCliConfirm.mockResolvedValue(true);
  };

  const doMockUserHasAccessToOriginOrg = () => {
    mockedGetUserHasAccessToOrg.mockResolvedValue(true);
  };

  beforeAll(() => {
    mockedAuthenticatedClient.mockImplementation(
      () =>
        ({
          getUserHasAccessToOrg: mockedGetUserHasAccessToOrg,
          getClient: mockedGetClient.mockImplementation(() =>
            Promise.resolve({
              vault: {list: mockedVaultList, import: mockedVaultImport},
            } as unknown as PlatformClient)
          ),
        } as Partial<AuthenticatedClient> as AuthenticatedClient)
    );
  });

  afterEach(() => {
    jest.restoreAllMocks();
  });

  describe('when there is no origin organization attached to the snapshot', () => {
    beforeEach(() => {
      mockedProject.mockImplementation(
        () =>
          ({
            getResourceManifest: () => null,
          } as Partial<Project> as Project)
      );
    });

    it('should return false', async () => {
      expect(
        await tryTransferFromOrganization({
          reporter: {missingVaultEntries: []} as unknown as SnapshotReporter,
          snapshot: {} as Snapshot,
          projectPath: 'somePathYay',
          cfg: {} as Configuration,
        })
      ).toBe(false);
    });
  });

  describe('when the user refuse to transfer', () => {
    beforeEach(() => {
      doMockProjectWithOriginOrg();
      mockedCliConfirm.mockResolvedValue(false);
    });

    it('should return false', async () => {
      expect(
        await tryTransferFromOrganization({
          reporter: {} as SnapshotReporter,
          snapshot: {} as Snapshot,
          projectPath: 'somePathYay',
          cfg: {} as Configuration,
        })
      ).toBe(false);
    });
  });

  describe('when the user does not have access to the origin organization', () => {
    beforeEach(() => {
      doMockProjectWithOriginOrg();
      doMockUserAnswerYes();
      mockedGetUserHasAccessToOrg.mockResolvedValue(false);
    });

    it('should warn the user about their lack of access', async () => {
      await tryTransferFromOrganization({
        reporter: {} as SnapshotReporter,
        snapshot: {} as Snapshot,
        projectPath: 'somePathYay',
        cfg: {} as Configuration,
      });

      expect(mockedCliWarn).toBeCalledTimes(1);
      expect(mockedCliWarn.mock.calls[0][0]).toMatchSnapshot();
    });

    it('should return false', async () => {
      expect(
        await tryTransferFromOrganization({
          reporter: {} as SnapshotReporter,
          snapshot: {} as Snapshot,
          projectPath: 'somePathYay',
          cfg: {} as Configuration,
        })
      ).toBe(false);
    });
  });

  describe('when all the prechecks are OK', () => {
    beforeEach(() => {
      doMockProjectWithOriginOrg();
      doMockUserAnswerYes();
      doMockUserHasAccessToOriginOrg();
    });

    it('should get all the vault entries from the origin organization', async () => {
      const fakeNumberOfPages = 4;
      mockedVaultList.mockResolvedValue({
        totalPages: fakeNumberOfPages,
        items: [],
      });

      await tryTransferFromOrganization({
        reporter: {missingVaultEntries: []} as unknown as SnapshotReporter,
        snapshot: {} as Snapshot,
        projectPath: 'somePathYay',
        cfg: {} as Configuration,
      });

      expect(mockedVaultList).toBeCalledTimes(4);
      for (let i = 0; i < fakeNumberOfPages; i++) {
        expect(mockedVaultList.mock.calls[i][0]).toMatchObject({page: i});
      }
    });

    describe('when the origin org does not have all the vault entries missing from the target org', () => {
      const fakeOriginVaultEntries = ['foo', 'bar'];
      const fakeMissingVaultEntriesOnTarget = ['baz', 'bar'];
      beforeAll(() => {
        mockedVaultList.mockResolvedValue({
          totalPages: 1,
          items: fakeOriginVaultEntries.map((entryId) => ({key: entryId})),
        });
      });

      it('should warn with a `SnapshotMissingVaultEntriesFromOriginError`', async () => {
        await tryTransferFromOrganization({
          reporter: {
            missingVaultEntries: fakeMissingVaultEntriesOnTarget.map(
              (entryId) => ({vaultEntryId: entryId})
            ),
          } as unknown as SnapshotReporter,
          snapshot: {targetId: 'someTargetId'} as Snapshot,
          projectPath: 'somePathYay',
          cfg: {} as Configuration,
        });
        expect(
          mockedSnapshotMissingVaultEntriesFromOriginError
        ).toBeCalledTimes(1);
        expect(mockedSnapshotMissingVaultEntriesFromOriginError).toBeCalledWith(
          'someOriginOrg',
          'someTargetId',
          ['baz']
        );
        expect(mockedCliWarn).toBeCalledTimes(1);
        expect(mockedCliWarn.mock.calls[0][0]).toBeInstanceOf(
          SnapshotMissingVaultEntriesFromOriginError
        );
      });

      it('should return false', async () => {
        expect(
          await tryTransferFromOrganization({
            reporter: {
              missingVaultEntries: fakeMissingVaultEntriesOnTarget.map(
                (entryId) => ({vaultEntryId: entryId})
              ),
            } as unknown as SnapshotReporter,
            snapshot: {targetId: 'someTargetId'} as Snapshot,
            projectPath: 'somePathYay',
            cfg: {} as Configuration,
          })
        ).toBe(false);
      });
    });

    describe('when the origin org does have all the vault entries from the target org', () => {
      const fakeOriginVaultEntries = ['foo', 'bar', 'baz'];
      const fakeMissingVaultEntriesOnTarget = ['baz', 'bar'];
      beforeAll(() => {
        mockedVaultList.mockResolvedValue({
          totalPages: 1,
          items: fakeOriginVaultEntries.map((entryId) => ({key: entryId})),
        });
      });

      it('should try to transfer the vault entries from the origin org to the destination org', async () => {
        await tryTransferFromOrganization({
          reporter: {
            missingVaultEntries: fakeMissingVaultEntriesOnTarget.map(
              (entryId) => ({vaultEntryId: entryId})
            ),
          } as unknown as SnapshotReporter,
          snapshot: {
            id: 'someSnapshotId',
            targetId: 'someTargetId',
          } as Snapshot,
          projectPath: 'somePathYay',
          cfg: {} as Configuration,
        });
        expect(mockedGetClient).toBeCalledTimes(2);
        expect(mockedGetClient).toHaveBeenNthCalledWith(
          1,
          expect.objectContaining({
            organization: 'someOriginOrg',
          })
        );
        expect(mockedGetClient).toHaveBeenNthCalledWith(
          2,
          expect.objectContaining({
            organization: 'someTargetId',
          })
        );
        expect(mockedVaultImport).toBeCalledTimes(1);
        expect(mockedVaultImport).toBeCalledWith(
          'someSnapshotId',
          'someOriginOrg',
          VaultFetchStrategy.onlyMissing
        );
      });

      describe('when the transfer fails', () => {
        beforeEach(() => {
          mockedVaultImport.mockRejectedValueOnce('ohno!');
        });

        it('should warn the user and bubble up the error from the backend', async () => {
          await tryTransferFromOrganization({
            reporter: {
              missingVaultEntries: fakeMissingVaultEntriesOnTarget.map(
                (entryId) => ({vaultEntryId: entryId})
              ),
            } as unknown as SnapshotReporter,
            snapshot: {} as Snapshot,
            projectPath: 'somePathYay',
            cfg: {} as Configuration,
          });
          expect(mockedCliWarn).toBeCalledTimes(2);
          expect(mockedCliWarn).toHaveBeenNthCalledWith(2, 'ohno!');
        });

        it('should return false', async () => {
          expect(
            await tryTransferFromOrganization({
              reporter: {
                missingVaultEntries: fakeMissingVaultEntriesOnTarget.map(
                  (entryId) => ({vaultEntryId: entryId})
                ),
              } as unknown as SnapshotReporter,
              snapshot: {} as Snapshot,
              projectPath: 'somePathYay',
              cfg: {} as Configuration,
            })
          ).toBe(false);
        });
      });

      describe('when the transfer is succesful', () => {
        it('should return true', async () => {
          expect(
            await tryTransferFromOrganization({
              reporter: {
                missingVaultEntries: fakeMissingVaultEntriesOnTarget.map(
                  (entryId) => ({vaultEntryId: entryId})
                ),
              } as unknown as SnapshotReporter,
              snapshot: {} as Snapshot,
              projectPath: 'somePathYay',
              cfg: {} as Configuration,
            })
          ).toBe(true);
        });
      });
    });
  });
});
