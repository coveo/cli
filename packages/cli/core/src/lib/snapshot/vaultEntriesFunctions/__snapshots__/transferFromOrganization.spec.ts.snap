// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#tryTransferFromOrganization when the user does not have access to the origin organization should warn the user about their lack of access 1`] = `
"We mapped this snapshot to someOriginOrg.
If you want to transfer the vault entries from someOriginOrg to undefined,
authenticate with an account that has access to both organizations and then try again."
`;
